hw/formal/tools/jaspergold/parse-fpv-report.py

```diff
@@ -0,0 +1,217 @@
+
+import argparse
+import logging as log
+import re
+from pathlib import Path
+from collections import OrderedDict
+import sys
+
+import hjson
+
+
+def extract_messages(str_buffer, patterns):
+    '''Extract messages matching patterns from str_buffer as a dictionary.
+
+    The patterns argument is a list of pairs, (key, pattern). Each pattern is a regex
+    and all matches in str_buffer are stored in a dictionary under the paired key.
+    '''
+    results = OrderedDict()
+    for key, pattern in patterns:
+        if results.get(key) is not None:
+            results[key] += re.findall(pattern, str_buffer, flags=re.MULTILINE)
+        else:
+            results[key] = re.findall(pattern, str_buffer, flags=re.MULTILINE)
+
+    return results
+
+
+def extract_messages_count(str_buffer, patterns):
+    '''Extract messages matching patterns from full_file as a dictionary.
+
+    The patterns argument is a list of pairs, (key, pattern). Each pattern is a regex
+    and the total count of all matches in str_buffer are stored in a dictionary under
+    the paired key.
+    '''
+    results = OrderedDict()
+    for key, pattern in patterns:
+        if results.get(key) is not None:
+            results[key] += len(re.findall(pattern, str_buffer, flags=re.MULTILINE))
+        else:
+            results[key] = len(re.findall(pattern, str_buffer, flags=re.MULTILINE))
+
+    return results
+
+
+def parse_fpv_message(str_buffer):
+    '''Parse error, warnings, and failed properties from the log file'''
+    err_warn_patterns = [("errors", r"^ERROR: .*"),
+                         ("errors", r"^\[ERROR.*"),
+                         ("warnings", r"^WARNING: .*"),
+                         ("warnings", r"^\[WARN.*"),
+                         ("cex", r"^\[\d+\]\s+(\S*)\s+cex.*"),
+                         ("undetermined", r"^\[\d+\]\s+(\S*)\s+undetermined.*"),
+                         ("unreachable", r"^\[\d+\]\s+(\S*)\s+unreachable.*")]
+    return extract_messages(str_buffer, err_warn_patterns)
+
+
+def parse_fpv_summary(str_buffer):
+    '''Count errors, warnings, and property status from the log file'''
+    message_patterns = [("errors", r"^ERROR: .*"),
+                        ("errors", r"^\[ERROR.*"),
+                        ("warnings", r"^WARNING: .*"),
+                        ("warnings", r"^\[WARN.*"),
+                        ("proven", r"^\[\d+\].*proven.*"),
+                        ("cex", r"^\[\d+\].*cex.*"),
+                        ("covered", r"^\[\d+\].*covered.*"),
+                        ("undetermined", r"^\[\d+\].*undetermined.*"),
+                        ("unreachable", r"^\[\d+\].*unreachable.*")]
+    fpv_summary = extract_messages_count(str_buffer, message_patterns)
+
+    total_msg = fpv_summary["proven"] + fpv_summary["cex"] + fpv_summary["undetermined"]
+    if total_msg > 0:
+        pass_rate = (fpv_summary["proven"] + fpv_summary["undetermined"]) / total_msg * 100
+    else:
+        pass_rate = 0
```

Comment:
I wonder whether it makes sense to factor this out into a function?
```python
def format_percentage(good, bad):
    denom = good + bad
    pc = denom ? 100 * good / denom : 0
    return '{0:.2f} %'.format(round(pc, 2))
```

then you get

```python
fpv_summary['pass_rate'] = format_percentage(fpv_summary["proven"] + fpv_summary["cex"],
                                             fpv_summary["undetermined"])
fpv_summary['cov_rate'] = format_percentage(fpv_summary["covered"], fpv_summary["unreachable"])
```

Reply:
You are completely right about the ternary operator. Not sure what I was thinking! All looks good now :-)